File: ambeth/Ambeth.Util/ambeth/typeinfo/TypeInfoItemUtil.cs
using De.Osthus.Ambeth.Collections;
using System;
using System.Collections;
using System.Collections.Generic;

namespace De.Osthus.Ambeth.Typeinfo
{
    public sealed class TypeInfoItemUtil
    {
        public static Type GetElementTypeUsingReflection(Type type, Type genericType_onlyUsedInJava)
        {
			if (type.HasElementType)
			{
				return type.GetElementType();
			}
			else if (!type.IsGenericType)
			{
				return type;
			}
			Type genericTypeDef = type.GetGenericTypeDefinition();
			if (typeof(IList<>).IsAssignableFrom(genericTypeDef))
			{
				return type.GetGenericArguments()[0];
			}
			if (typeof(IMap<,>).IsAssignableFrom(genericTypeDef))
			{
				return GetElementTypeUsingReflection(type.GetGenericArguments()[1], null);
			}
            return type;
        }

        private TypeInfoItemUtil()
        {
            // intended blank
        }
    }
}

File: jambeth/jambeth-ioc/src/main/java/de/osthus/ambeth/typeinfo/TypeInfoItemUtil.java
package de.osthus.ambeth.typeinfo;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.lang.reflect.TypeVariable;
import java.lang.reflect.WildcardType;
import java.util.Collection;

public final class TypeInfoItemUtil
{
	public static Class<?> getElementTypeUsingReflection(Class<?> propertyType, Type genericType)
	{
		if (propertyType.isArray())
		{
			return propertyType.getComponentType();
		}
		else if (Collection.class.isAssignableFrom(propertyType))
		{
			if (!(genericType instanceof ParameterizedType))
			{
				if (genericType instanceof Class)
				{
					return (Class<?>) genericType;
				}
				return Object.class;
			}
			ParameterizedType castedType = (ParameterizedType) genericType;
			Type[] actualTypeArguments = castedType.getActualTypeArguments();
			if (actualTypeArguments.length == 1)
			{
				Type actualType = actualTypeArguments[0];
				if (actualType instanceof ParameterizedType)
				{
					propertyType = (Class<?>) ((ParameterizedType) actualType).getRawType();
				}
				else if (actualType instanceof TypeVariable)
				{
					return Object.class;
				}
				else if (actualType instanceof WildcardType)
				{
					Type[] upperBounds = ((WildcardType) actualType).getUpperBounds();
					return getElementTypeUsingReflection(propertyType, upperBounds[0]);
				}
				else
				{
					propertyType = (Class<?>) actualType;
				}
			}
			else if (actualTypeArguments.length == 0)
			{
				// TODO This should be logged, but no logger is available
			}
			else
			{
				throw new IllegalArgumentException("Properties with more than one generic type are not supported");
			}
		}

		return propertyType;
	}

	private TypeInfoItemUtil()
	{
		// Intended blank
	}
}
